imap/jmap_mailbox.c

```diff
@@ -3686,6 +3686,31 @@ static void _mboxset_state_update_specialuse(struct mboxset_state *state,
         }
     }
 
+    /* Validate: no children for specialuse_nochildren roles */
+    if (config_getstring(IMAPOPT_SPECIALUSE_NOCHILDREN)) {
+        strarray_t *nochildren =
+            strarray_split(config_getstring(IMAPOPT_SPECIALUSE_NOCHILDREN),
+                NULL, STRARRAY_TRIM);
+
+        hash_iter *iter = hash_table_iter(state->specialuses_by_id);
+        while (hash_iter_next(iter)) {
+            const char* mbox_id = hash_iter_key(iter);
+            strarray_t *specialuses = hash_iter_val(iter);
+            if (strarray_size(specialuses) &&
+                    strarray_intersect(specialuses, nochildren)) {
+                if (hash_lookup(mbox_id, state->siblingnames_by_parent_id)) {
+                    buf_printf(&conflict_desc, "\nMailbox %s has specialuse %s "
+                            "which forbids it to have children", mbox_id,
+                            strarray_nth(specialuses, 0));
```

Comment:
Am I reading correctly here, that it will report the mailbox's first (`0`th) specialuse as being the cause of the conflict, regardless of which one actually conflicted?

I guess if a mailbox always has 0 or 1 specialuse flags, then that's correct, but I'm not sure that one can't have multiple.  Though I'm not sure one mailbox having multiple specialuses would make sense anyway.

I think maybe I'm getting confused by `specialuses` here being a `strarray_t`.  I guess maybe we use that, even though it can only have 0 or 1 value, for algorithmic ease elsewhere (such as being able to use `strarray_intersect()`)?

Reply:
The idea is when we got to this point we already validated that a mailbox must have zero or one specialuses (JMAP does not allow more than one per mailbox). That's why we use strarray_t here, its the working data of the previous validation step that asserts this condition.
What my initial code did not was to skip validating specialue_nochildren if the previous specialuse validation already determined the state to be invalid. I pushed a new commit to fix this.